divfree/TPZCompElHCurlNoGrads.cpp

```diff
@@ -0,0 +1,369 @@
+
+  TPZManVector<int64_t,nNodes> nodes(nNodes, 0);
+
+  for (auto iNode = 0; iNode < nNodes; iNode++){
+    nodes[iNode] = this->Reference()->NodeIndex(iNode);
+  }
+
+  
+  TPZManVector<int64_t, TSHAPE::NSides - nNodes>
+    firstH1ShapeFunc(TSHAPE::NSides - nNodes,0);
+  
+  //calculates the first shape function associated with each side of dim > 0
+  TPZManVector<int,TSHAPE::NSides-nNodes> sidesH1Ord(TSHAPE::NSides - nNodes,-1);
+  this->CalcH1ShapeOrders(sidesH1Ord);
+  firstH1ShapeFunc[0] = nNodes;
+  
+  for (int iSide = nNodes + 1; iSide < TSHAPE::NSides; iSide++) {
+    const int iCon = iSide - nNodes;
+    firstH1ShapeFunc[iCon] =
+      firstH1ShapeFunc[iCon - 1] +
+      TSHAPE::NConnectShapeF(iSide - 1, sidesH1Ord[iCon-1]);
+
+  }
+
+  auto &indexVecShape = data.fVecShapeIndex;
+
+
+  
+  
+  indexVecShape.Resize(unfiltnshape);
+  TPZVec<unsigned int> shapeCountVec(TSHAPE::NSides - nNodes, 0);
+  this->StaticIndexShapeToVec(indexVecShape, connectOrders,
+                              firstH1ShapeFunc,sidesH1Ord, shapeCountVec, nodes);
+
+  //setting the type of shape functions as vector shape functions
+  data.fShapeType = TPZMaterialData::EVecShape;
+}
+
+template<class TSHAPE>
+template<class TVar>
+void TPZCompElHCurlNoGrads<TSHAPE>::ComputeRequiredDataT(
+  TPZMaterialDataT<TVar> &data, TPZVec<REAL> &qsi)
+{
+  
+  {
+    const bool needsSol = data.fNeedsSol;
+    data.fNeedsSol = false;
+    TPZIntelGen<TSHAPE>::ComputeRequiredData(data,qsi);//in this method, Shape will be called
+    data.fNeedsSol = needsSol;
+  }
+
+  this->ComputeDeformedDirections(data);    
+  /******************************************************************************************************************
+   * at this point, we already have the basis functions on the deformed element, since we have data.phi,
+   * data.fVecShapeIndex and data.fDeformedDirections. Now it is time to compute the curl, which will be stored in
+   * data.curlphi.
+   *******************************************************************************************************************/
+  TPZFMatrix<REAL> phiHCurl;
+  this->ComputeShape(data, phiHCurl);
+  
+
+  constexpr auto dim{TSHAPE::Dimension};
+  data.curlphi.Redim(2*dim - 3 > 0 ? 2*dim - 3 : 1, this->NShapeF());
+  this->ComputeCurl<dim>(data);
+  
+  data.phi = phiHCurl;
+  if (data.fNeedsSol) {
+    this->ReallyComputeSolution(data);
+  }
+}
+
+template<class TSHAPE>
+template<class TVar>
+void TPZCompElHCurlNoGrads<TSHAPE>::ReallyComputeSolutionT(TPZMaterialDataT<TVar> &data)
+{
+    this->ComputeSolutionHCurlT(data.phi, data.curlphi,
+                         data.sol, data.curlsol);
+}
+
+
+template<class TSHAPE>
+int TPZCompElHCurlNoGrads<TSHAPE>::NConnectShapeF(int icon, int order) const
+{
+  const int side = icon + TSHAPE::NCornerNodes;
+#ifdef PZDEBUG
+  if (side < TSHAPE::NCornerNodes || side >= TSHAPE::NSides) {
+    DebugStop();
+  }
+#endif
+  if(order == 0) return 0;
```

Comment:
In HCurl, if the order is zero (constant function) then the number of shape functions should be 1
Or if order zero is not allowed, then DebugStop()

Reply:
Addressed in https://github.com/labmec/divfreebubbles/pull/1#discussion_r691489964

Reply:
But still you should call DebugStop()

Reply:
Resolved in 6a0a189